File: backend/app/models/payment.py
from datetime import datetime, timezone

from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer, Numeric, String
from sqlalchemy.orm import relationship

from app.database import Base


class Payment(Base):
    __tablename__ = "payments"

    id = Column(Integer, primary_key=True, index=True)
    shipment_id = Column(Integer, ForeignKey("shipments.id"), nullable=True, index=True)
    cod_id = Column(Integer, ForeignKey("cods.id"), nullable=True, index=True)
    customer_id = Column(Integer, ForeignKey("customers.id"), nullable=False, index=True)
    company_id = Column(Integer, ForeignKey("companies.id"), nullable=True, index=True)
    amount = Column(Numeric(10, 2), nullable=False)
    currency = Column(String, nullable=False, default="EGP")
    payment_method = Column(String, nullable=False)
    payment_status = Column(String, nullable=False)
    transaction_reference = Column(String, nullable=False, unique=True, index=True)
    paid_at = Column(DateTime, nullable=True)
    notes = Column(String, nullable=True)
    is_deleted = Column(Boolean, default=False, nullable=False)
    deleted_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc), nullable=False)

    shipment = relationship("Shipment", back_populates="payments")
    cod = relationship("COD", back_populates="payments")
    customer = relationship("Customer", back_populates="payments")

File: frontend/types/payment.ts
export type Payment = {
  id: number
  shipment_id: number | null
  cod_id: number | null
  customer_id: number
  company_id: number | null
  amount: number
  currency: string
  payment_method: string
  payment_status: string
  transaction_reference: string
  paid_at: string | null
  notes: string | null
  is_deleted: boolean
  deleted_at: string | null
  created_at: string
  updated_at: string
}

export type PaymentCreatePayload = {
  shipment_id?: number | null
  cod_id?: number | null
  customer_id: number
  amount: number
  currency?: string
  payment_method: string
  payment_status: string
  transaction_reference: string
  paid_at?: string | null
  notes?: string | null
}

export type PaymentUpdatePayload = PaymentCreatePayload
